### PR TITLE
CPO-16 - Use Hibernate Envers library for auditing table

### PR DESCRIPTION
**Pull request by sullivangareth**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/CPO-16

### Change description ###

Use an auditing table to track changes made to `case_payments_order` table

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (danlysiak) [inline src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:42]: We can delete this

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java
Comment (danlysiak) [inline src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java:6]: This whole class can be deleted - through the existing magic of what's currently in `master`, a postgres container will already start up and be automatically configured as the datasource

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:21]: Nice 👌

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java]: 👍

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java]: 👍

Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java]: Fixed in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/997e6c96e81b2bb04894eeb6e64f23be1a43b5c9

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java]: Fixed in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/997e6c96e81b2bb04894eeb6e64f23be1a43b5c9

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/PostgresqlContainer.java]: The magic is in the `application-itest.yaml` file, specifically the `datasource.url`

 `jdbc:tc:postgresql:11:///databasename`
 
 Note there is no `host` or `port` specified - these are not needed and the `testcontainers` library with automagically creatae an instance of the container
 
 More details [heee at testcontainters.org](https://www.testcontainers.org/modules/databases/jdbc/)

File: build.gradle
Comment (smathangi) [inline build.gradle:278]: It's not related to your PR. But could you remove unwanted testContainers dependencies from Gradle please. 
https://github.com/hmcts/cpo-case-payment-orders-api/blob/master/build.gradle#L295-L296

**Review by mario-paniccia — DISMISSED** [review]
👍

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (smathangi) [inline src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:30]: Can't we use @DataJpaTest or something else instead of loading whole Spring Boot app. ?

File: build.gradle
Comment (sullivangareth) [reply build.gradle:278]: Why are they unwanted ?

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (mario-paniccia) [inline src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:72]: this API is a bit complicated to understand. To make unit tests more understandable, consider "hiding" this complexity behind a private method with a nice explanatory name

Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:72]: It is a horrible API for sure - will do

**Review by mario-paniccia — COMMENTED** [review]
one comment to improve tests readability

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:38]: Why are we moving caseTypeId from the entity , is it no longer needed ?

File: src/main/resources/db/migration/V20210322_16.2__CPO-16-CasePaymentOrdersAUD.sql
Comment (smathangi) [inline src/main/resources/db/migration/V20210322_16.2__CPO-16-CasePaymentOrdersAUD.sql:23]: do we ever change created columns ? If not we can exclude from auditing (see @NotAudited)

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:38]: Was recently removed from the Data Model

https://tools.hmcts.net/confluence/display/RCCD/Case+Payment+Orders#CasePaymentOrders-DataModel

This work is also being done as a separate ticket (https://tools.hmcts.net/jira/browse/CPO-21), so likely that I will wait for that to be merged before merging this PR

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:30]: We could I guess, but I do like the simplicity of extending from the `BaseTest` for consistency, and for time being the app is small enough that I don't think it would have much impact on performance of tests?

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:21]: Any business requirement for "querying revisions of entity that modified given property" ? otherwise 'withModifiedFlag' flag comes with some cost.

Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:21]: https://tools.hmcts.net/jira/browse/CPO-16 states

`Note that we want to use @Audited( withModifiedFlag=true ) to enable tracking which properties of audited entities were modified in a given revision.`

File: build.gradle
Comment (smathangi) [reply build.gradle:278]: we are spinning up container using datasource url (which requires postgres-container dependency only, that pulls automatically core).

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:21]: @mario-paniccia  ^^

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (smathangi) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:30]: I thought it's DAO layer specific test. nw if setup is a bit complicated.

File: build.gradle
Comment (sullivangareth) [reply build.gradle:278]: Ah right, ok - fixed in  6220d88

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java]: refactored in commit https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/2feddd456da4c714ca6caed9d8207ee8bc40cf46

File: src/main/resources/db/migration/V20210322_16.2__CPO-16-CasePaymentOrdersAUD.sql
Comment (sullivangareth) [reply src/main/resources/db/migration/V20210322_16.2__CPO-16-CasePaymentOrdersAUD.sql:22]: `created_by_mod` & `created_timestamp_mod` removed in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/2feddd456da4c714ca6caed9d8207ee8bc40cf46

Comment (mario-paniccia) [reply src/main/resources/db/migration/V20210322_16.2__CPO-16-CasePaymentOrdersAUD.sql:22]: @sullivangareth let's remember to update Nigel's wiki then please

Comment (sullivangareth) [reply src/main/resources/db/migration/V20210322_16.2__CPO-16-CasePaymentOrdersAUD.sql]: Have reinstated auditing on all fields for now, in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/919ae706dbe69f88ba6404b0bf800b8d624f37d1

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java]: CPO-21 has been merged to master, and pulled into this ticket in commit https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/aee9fd201481fa9e4808e08cc2a1bb2ba282fa78

File: src/main/resources/db/migration/V20210322_16__CPO-16-CasePaymentOrdersAUD.sql
Comment (danlysiak) [inline src/main/resources/db/migration/V20210322_16__CPO-16-CasePaymentOrdersAUD.sql:24]: We should update the scope doc to include `created_timestamp_mod` if required

**Review by danlysiak — APPROVED** [review]
Approved with one comment

File: src/main/resources/db/migration/V20210322_16__CPO-16-CasePaymentOrdersAUD.sql
Comment (sullivangareth) [reply src/main/resources/db/migration/V20210322_16__CPO-16-CasePaymentOrdersAUD.sql]: Fixed in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/450c4f00d76e0d3016b2e00af0cef519922b088b

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (mario-paniccia) [inline src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:116]: to make the test more readable, it would be good to have also this extracted in a private method with a nice explanatory name. 
Since it takes a while to understand what this is doing, given the complexity of the API

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (mario-paniccia) [inline src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:103]:
```suggestion
    void testAuditOfDeletingCasePaymentOrder() {
```
this name is not consistent with the others. We are not testing the creation right? It's been tested in the dedicated test method already

File: src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/reform/cpo/data/CasePaymentOrderEntity.java:22]: No business requirement for that, it's true. But given that I didn't expect the amount of changes to CPOs to be that big I thought it won't hurt to use it. 
But if we feel we want to apply YAGNI, ok for me to remove it. As you prefer

**Review by mario-paniccia — APPROVED** [review]
👍 
some minor comments

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java:123]: Fixed in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/448f8f49d429b7c4e7617d980aef2ce4bc8e100e

File: src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/cpo/repository/CasePaymentOrdersAuditTest.java]: Fixed in https://github.com/hmcts/cpo-case-payment-orders-api/pull/29/commits/448f8f49d429b7c4e7617d980aef2ce4bc8e100e

**Pull request merged**
